zerver/actions/message_send.py

```diff
@@ -965,6 +967,36 @@ def do_send_messages(
                         visibility_policy=new_visibility_policy,
                     )
 
+            # Set the visibility_policy of the users mentioned in the message
+            # to "FOLLOWED" if "follow_topics_where_mentioned" is "True".
+            user_mentions = send_request.rendering_result.mentions_user_ids
+
+            for user_id in user_mentions:
+                user_profile = get_user_profile_by_id(user_id)
```

Comment:
If a message has `@all` wildcard mention and the stream has 20k subscribers, this will cost 20k x 4 database queries..., right? We need to improve this.

Reply:
This setting doesn't work for wildcard mentions. Should it?

Reply:
Oh, right. We don't need to consider wildcard mentions here.

Reply:
In `build_message_send_dict`
```
    # For single user as well as user group mentions, we set the `mentioned`
    # flag on `UserMessage`
    for group_id in rendering_result.mentions_user_group_ids:
        members = mention_data.get_group_members(group_id)
        rendering_result.mentions_user_ids.update(members)
 ```
 
 So, `mentions_user_ids` contains multiple ids in the case of group mention; hence db queries in loop.
 
 We can use `bulk_do_set_user_topic_visibility_policy`. Also we need to find a way to store these user_profiles
(get_user_profile_by_id(user_id)) early in send message code path if fetched already. The ultimate goal is to minimize the db queries in this codepath as this can slow down things.

Reply:
I think it will be better to store mentioned user_ids in ```SendMessageRequest``` object. We can check if the mentioned user is subscribed to the stream using already fetched recipient info in ```build_message_send_dict``` . This will decrease 1 query but there will still be 5 extra queries: 1 to fetch the user_profiles(as they are not fetched anywhere else) + 1 to check if the topic is already followed + 3 to follow the topic, per mentioned user in the message.

Reply:
Hm, I think that plan would work.

Reply:
I have used ```send_request.active_user_ids``` to decrease the query count by 1. Also, used ```bulk_do_set_user_topic_visibility_policy``` instead of ```do_set_user_topic_visibility_policy```.

Reply:
I think we can't omit the query for checking the visibility policy of the user. If we can get user_profile directly from the mentions instead of user_ids, then we can decrease 1 query but I am not sure about that. Should I start the discussion then?

Reply:
So, a better plan would be to fetch user profiles in one query Userprofile.objects.filter(id_in=user_mentions); instead of fetching them in loop. Similar query on UserTopic to fetch the required data in one query instead of individual queries.

We might need to use some extra data structures to manage the fetched data properly. some kind of `user_profile` to `user_topic` map. then iterate. Sounds good?

But still please confirm on CZO, if we plan to include group mentioned and wildcard mention or not; to avoid rework.

Reply:
OK, I have started a [discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/.22automatically.20follow.20topics.20where.20I'm.20mentioned.22.20setting) on CZO.

Reply:
Where is this discussion? The link doesn't work, and #design is the right stream for discussing questions about what user experience we want.

Reply:
https://chat.zulip.org/#narrow/stream/9-issues/topic/Mentioned.20in.20Recent.20conversations/near/1678211

Reply:
I have used ```Userprofile.objects.filter(id__in=user_mentions)``` to fetch ```user_profiles``` in just 1 query and am trying to use 
```
user_topic = UserTopic.objects.filter(
        user_profile__in=user_profiles,
        stream_id=send_request.stream.id,
        topic_name__iexact=send_request.message.topic_name(),
 )
```
This can be used to check ```visibility_policy``` of fetched users. But if "UserTopic.DoesNotExist" error arise(that topic will not be in the query set), we will change the visibility_policy anyway.

@prakhar1144 Does this sound good?

Reply:
I have done necessary changes in ```message_send.py``` to reduce db queries. @prakhar1144 Can you please review that?